nav2_costmap_2d/test/unit/speed_filter_test.cpp

```diff
@@ -0,0 +1,739 @@
+  void rePublishMask();
+  void createSpeedFilter(const std::string & global_frame);
+  void createTFBroadcaster(const std::string & mask_frame, const std::string & global_frame);
+
+  // Test methods
+  void testFullMask(
+    uint8_t type, double base, double multiplier,
+    double tr_x, double tr_y);
+  void testSimpleMask(
+    uint8_t type, double base, double multiplier,
+    double tr_x, double tr_y);
+  void testOutOfMask(uint8_t type, double base, double multiplier);
+  void testIncorrectPercentage(uint8_t type, double base, double multiplier);
+
+  void reset();
+
+  std::shared_ptr<nav2_costmap_2d::SpeedFilter> speed_filter_;
+  std::shared_ptr<nav2_costmap_2d::Costmap2D> master_grid_;
+
+private:
+  void waitSome(const std::chrono::nanoseconds & duration);
+  void verifySpeedLimit(
+    uint8_t type, double base, double multiplier,
+    unsigned int x, unsigned int y,
+    nav2_msgs::msg::SpeedLimit::SharedPtr speed_limit);
+  nav2_msgs::msg::SpeedLimit::SharedPtr getSpeedLimit();
+  nav2_msgs::msg::SpeedLimit::SharedPtr waitSpeedLimit();
+
+  const unsigned int width_ = 10;
+  const unsigned int height_ = 11;
+  const double resolution_ = 1.0;
+
+  nav2_util::LifecycleNode::SharedPtr node_;
+
+  std::shared_ptr<tf2_ros::Buffer> tf_buffer_;
+  std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
+  std::shared_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;
+  std::unique_ptr<geometry_msgs::msg::TransformStamped> transform_;
+
+  std::shared_ptr<TestMask> mask_;
+
+  std::shared_ptr<InfoPublisher> info_publisher_;
+  std::shared_ptr<MaskPublisher> mask_publisher_;
+  std::shared_ptr<SpeedLimitSubscriber> speed_limit_subscriber_;
+};
+
+void TestNode::createMaps(const std::string & mask_frame)
+{
+  // Make map and mask put as follows:
+  //  master_grid     (12,13)
+  //    *----------------*
+  //    |                |
+  //    |  mask  (10,11) |
+  //    |   *-------*    |
+  //    |   |///////|    |
+  //    |   |///////|    |
+  //    |   |///////|    |
+  //    |   *-------*    |
+  //    | (0,0)          |
+  //    |                |
+  //    *----------------*
+  // (-2,-2)
+
+  // Create master_grid_
+  master_grid_ = std::make_shared<nav2_costmap_2d::Costmap2D>(
+    width_ + 4, height_ + 4, resolution_, -2.0, -2.0, nav2_costmap_2d::FREE_SPACE);
+
+  // Create mask_
+  mask_ = std::make_shared<TestMask>(width_, height_, resolution_, mask_frame);
+}
+
+void TestNode::publishMaps(uint8_t type, double base, double multiplier)
+{
+  info_publisher_ = std::make_shared<InfoPublisher>(type, base, multiplier);
+  mask_publisher_ = std::make_shared<MaskPublisher>(*mask_);
+}
+
+void TestNode::rePublishInfo(uint8_t type, double base, double multiplier)
+{
+  info_publisher_.reset();
+  info_publisher_ = std::make_shared<InfoPublisher>(type, base, multiplier);
+  // Allow both CostmapFilterInfo and filter mask subscribers
+  // to receive a new message
+  waitSome(500ms);
```

Comment:
Is 500ms necessary? Can this be dropped down? These `wait`s will add non-trivial time to the tests, I see a bunch of them.

Reply:
`waitSome(time)` routine not only waits, but also spins all nodes in this test. I believe, this can not be removed because of some time is required to send-receive the message to/from topic (e.g. in this case it `CostmapFilterInfo` message) and `spin_some(all_nodes)` to handle this message.

However, I checked that initial timeouts are excessive to settle down. So, I've decreased them for [keepout](https://github.com/ros-planning/navigation2/commit/34e694dbee7af9f805d2a7038b292466e44dc104#diff-be511f9657acb8524951279503a1173254f7af6cfd32e607e7bd1572a2aa638d) and [speed](https://github.com/ros-planning/navigation2/commit/34e694dbee7af9f805d2a7038b292466e44dc104#diff-b4571d8a4a9cb6222160b2ef3d874689b3eaf5e86a21928869f3b02a90c6b44f) filters unit tests. That's what i got:
```
--- keepout_filter_test time ---
initial:        4.161 sec
optimized:      1.613 sec

--- speed_filter_test time ---
initial:        27.185 sec
optimized:      5.857 sec
```
Testing was stable in infinite loop for both tests running ~15min on my PC (with stress test of CPU and system HDD going in parallel).
So, hope this will improve the situation.

Reply:
27 -> 5 seconds is a good improvement, thanks